### PR TITLE
JIT: (x u>> cns) -> [0..(x's max value >> cns)]

### PR DESCRIPTION
**Pull request by EgorBo**
Optimizes bounds check for https://github.com/dotnet/runtime/issues/109899

### CODE REVIEW

File: src/coreclr/jit/rangecheck.cpp [hunk 1]
```diff
@@ -1049,6 +1049,29 @@ Range RangeCheck::ComputeRangeForBinOp(BasicBlock* block, GenTreeOp* binop, bool
...
+            {
+                return Range(Limit(Limit::keConstant, 0));
+            }
+
+            // Calculate max possible value of op1, e.g. UINT_MAX for TYP_INT/TYP_UINT
+            uint64_t maxValue = 0;
+            for (int bit = 0; bit < op1Width; bit++)
+            {
+                maxValue |= 1ULL << bit;
+            }
```
Comment (teo-tsirpanis) [inline src/coreclr/jit/rangecheck.cpp:1072]:
```suggestion
            uint64_t maxValue = (1ULL << op1Width) - 1;
```

Comment (EgorBo) [reply src/coreclr/jit/rangecheck.cpp]: that's better, thanks!

**Comment by EgorBo** [comment]
@EgorBot -codesafety

File: src/coreclr/jit/rangecheck.cpp [hunk 2]
```diff
@@ -1049,6 +1049,25 @@ Range RangeCheck::ComputeRangeForBinOp(BasicBlock* block, GenTreeOp* binop, bool
...
+                return Range(Limit::keUnknown);
+            }
+
+            int op1Width = (int)(genTypeSize(op1) * BITS_PER_BYTE);
+            if (shiftBy >= op1Width)
+            {
+                return Range(Limit(Limit::keConstant, 0));
+            }
+
+            // Calculate max possible value of op1, e.g. UINT_MAX for TYP_INT/TYP_UINT
```
Comment (AndyAyersMS) [inline src/coreclr/jit/rangecheck.cpp:1067]: Was wondering if we could use `IntegralRange` here but that seems a bit clunky.

**Review by AndyAyersMS — APPROVED** [review]

File: src/coreclr/jit/rangecheck.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/jit/rangecheck.cpp:1067]: Initially, I wanted to call `RangeCheck::ComputeRange` on op1 to get even more precise range, but it had no diffs

**Pull request merged**